agent/exec/containerd/adapter.go

```diff
@@ -0,0 +1,682 @@
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	contentservice "github.com/containerd/containerd/services/content"
+	imagesservice "github.com/containerd/containerd/services/images"
+	"github.com/containerd/fifo"
+	dockermount "github.com/docker/docker/pkg/mount"
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/api/naming"
+	"github.com/docker/swarmkit/log"
+	protobuf "github.com/gogo/protobuf/types"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+var (
+	mountPropagationReverseMap = map[api.Mount_BindOptions_MountPropagation]string{
+		api.MountPropagationPrivate:  "private",
+		api.MountPropagationRPrivate: "rprivate",
+		api.MountPropagationShared:   "shared",
+		api.MountPropagationRShared:  "rshared",
+		api.MountPropagationRSlave:   "slave",
+		api.MountPropagationSlave:    "rslave",
+	}
+)
+
+// containerAdapter conducts remote operations for a container. All calls
+// are mostly naked calls to the client API, seeded with information from
+// containerConfig.
+type containerAdapter struct {
+	conn              *grpc.ClientConn
+	client            execution.ContainerServiceClient
+	container         *api.ContainerSpec
+	task              *api.Task
+	secrets           exec.SecretGetter
+	dir               string
+	resolvedImageName string
+	deleteResponse    *execution.DeleteResponse
+}
+
+func newContainerAdapter(conn *grpc.ClientConn, containerDir string, task *api.Task, secrets exec.SecretGetter) (*containerAdapter, error) {
+	container := task.Spec.GetContainer()
+	if container == nil {
+		return nil, exec.ErrRuntimeUnsupported
+	}
+
+	dir := filepath.Join(containerDir, task.ID)
+
+	return &containerAdapter{
+		conn:      conn,
+		client:    execution.NewContainerServiceClient(conn),
+		container: container,
+		task:      task,
+		secrets:   secrets,
+		dir:       dir,
+	}, nil
+}
+
+func (c *containerAdapter) applyLayer(ctx context.Context, cs content.Store, rootfs string, layer digest.Digest) error {
+	blob, err := cs.Reader(ctx, layer)
+	if err != nil {
+		return err
+	}
+
+	rd, err := compression.DecompressStream(blob)
+	if err != nil {
+		return err
+	}
+
+	_, err = archive.Apply(ctx, rootfs, rd)
+
+	blob.Close()
+	return err
+}
+
+// github.com/containerd/containerd cmd/ctr/utils.go, dropped stdin handling
+func prepareStdio(stdout, stderr string, console bool) (*sync.WaitGroup, error) {
+	var wg sync.WaitGroup
+	ctx := context.Background()
+
+	f, err := fifo.OpenFifo(ctx, stdout, syscall.O_RDONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700)
+	if err != nil {
+		return nil, err
+	}
+	defer func(c io.Closer) {
+		if err != nil {
```

Comment:
It's a good idea to use a named return. Otherwise, if something returns an error and the variable is named differently or is shaddowed inside a deeper scope, this would miss the error.